cprover_bindings/src/goto_program/contract.rs

```diff
@@ -12,7 +12,7 @@ use std::fmt::Debug;
 /// Represents a contract on a function, loop, etc.
 #[derive(Clone, Debug)]
 pub enum Contract {
-    FunctionContract { ensures: Vec<Spec>, requires: Vec<Spec> },
+    FunctionContract { ensures: Vec<Spec>, requires: Vec<Spec>, assigns: Vec<Spec> },
```

Comment:
Is the same `Spec` type suitable for `assigns` clauses? What would an `assigns` clause look like in Rust? Would it refer to expressions or just symbols?

Reply:
Yeah, `assigns: Vec<Spec>` is too general. An assigns clause `__CPROVER_assigns(x)` gets wrapped into a lambda expression on the CBMC side. So, it will look something like - (|function_return_type, function_arguments| x). 

On the Rust side, it would be `kani::assigns(x)` where x is an `Expr` with `ExprValue` being a symbol. 

I've updated the code to make sure that the only way we can construct a `FunctionContract` type is through the constructor. In the constructor, in checking to make sure that each `assigns(x)` clause satisfies `x.is_symbol()`. 

Does this sound good?

Reply:
Sounds good.